src/section/shared_state.rs

```diff
@@ -173,15 +179,32 @@ impl SharedState {
         // FIXME: we should perform these checks before voting, but once the vote accumulates we
         // must obey it:
         if !self.our_prefix().matches(p2p_node.name()) {
-            trace!("not adding node {} - not matching our prefix", p2p_node);
+            trace!(
+                "not adding node {} - not matching our prefix",
+                p2p_node.name()
+            );
             return false;
         }
 
         if self.our_members.contains(p2p_node.name()) {
-            trace!("not adding node {} - already a member", p2p_node);
+            trace!("not adding node {} - already a member", p2p_node.name());
             return false;
         }
 
+        // The section public key of the proof shall be trusted by us.
+        match self.our_history.check_trust(iter::once(&proof.public_key)) {
```

Comment:
So this checks that the proof key is known to us, but we should also check the proof itself is valid by verifying the signature. This applies to other places too.

Reply:
Also, `check_trust` isn't supposed to be used like this, although it might still happen to work. What `check_trust` does, is: given a collection of keys that we already trust, return whether the proof chain itself is trusted. So you would use it to verify the proof chain itself, not the keys. But we don't need to verify the proof chain here, we can assume it is valid already. What you should use is just `self.our_history.has_key(&proof.public_key)`.

Reply:
> but we should also check the proof itself is valid by verifying the signature. This applies to other places too

I thought this has been carried out during message accumulation already?

> What you should use is just self.our_history.has_key(&proof.public_key)

Ok, will change to this

Reply:
> I thought this has been carried out during message accumulation already?

Yeah that's true. Hm. Then maybe it's not necessary to do those checks here. Only where we know the input might come from untrusted sources (so for example directly from a received message, not accumulated).

Reply:
> so for example directly from a received message, not accumulated

yes.
So, basically, if the current mutation function has a `proof` passed it, it indicates it's `accumulated`, wo we don't need to check the signature validity anymore.
If no `proof` provided, then it maybe a direct message, then shall be checked. (however, maybe due to the serialisation/derialization, the checking place shall be in the calling place, not here.)